src/app/linux.rs

```diff
@@ -1,20 +1,304 @@
+    stats: &'a HashMap<&str, SnapshotParseData>,
+) -> Option<Snapshot> {
+    use SnapshotParseData::{Name, Number};
+    let get_name = |data: &'a SnapshotParseData| match data {
+        Name(name) => Some(name),
+        _ => None,
+    };
+    let get_number = |data: &'a SnapshotParseData| match data {
+        Number(n) => Some(*n),
+        _ => None,
+    };
+    let name = stats.get("dataset_name").and_then(get_name)?.to_string();
+    let nunlinked = stats.get("nunlinked").and_then(get_number)?;
+    let nunlinks = stats.get("nunlinks").and_then(get_number)?;
+    let nread = stats.get("nread").and_then(get_number)?;
+    let reads = stats.get("reads").and_then(get_number)?;
+    let nwritten = stats.get("nwritten").and_then(get_number)?;
+    let writes = stats.get("writes").and_then(get_number)?;
+    Some(Snapshot {
+        name,
+        nunlinked,
+        nunlinks,
+        nread,
+        reads,
+        nwritten,
+        writes,
+    })
+}
+
+#[derive(Debug, PartialEq, Eq)]
+enum ZTopError {
+    PoolDoesNotExist { pool: String },
+    SnapshotParseError { src: String },
+}
+
+impl Error for ZTopError {}
+
+impl fmt::Display for ZTopError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            ZTopError::SnapshotParseError { src } => {
+                write!(f, "Failed to parse Snapshot from string:\n {}", src)
+            }
+            ZTopError::PoolDoesNotExist { pool } => {
+                write!(f, "ZFS pool does not exist: {}", pool)
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+enum SnapshotParseData<'a> {
+    Name(&'a str),
+    Number(u64),
+}
+
+fn parse_snapshot(s: &str) -> Result<Snapshot, ZTopError> {
+    use SnapshotParseData::{Name, Number};
+    let mut stats = HashMap::new();
+    for row in s.split('\n') {
+        let fields: Vec<_> = row.split_whitespace().collect();
+        match fields[..] {
+            [name, _, data] if name != "name" => {
+                let data =
+                    data.parse::<u64>().map_or_else(|_| Name(data), Number);
+                stats.insert(name, data);
+            }
+            _ => {}
+        }
+    }
+    if let Some(snap) = snapshot_from_hash_map(&stats) {
+        Ok(snap)
+    } else {
+        Err(ZTopError::SnapshotParseError { src: s.to_owned() })
+    }
+}
+
+#[cfg(test)]
+mod t {
+    mod parsing {
+        use super::super::*;
+
+        const OBJSET: [&str; 9] = [
+            "26 1 0x01 7 2160 15278395714 17176723400350",
+            "name                            type data",
+            "dataset_name                    7    tank",
+            "writes                          4    14",
+            "nwritten                        4    256",
+            "reads                           4    100",
+            "nread                           4    1024",
+            "nunlinks                        4    1",
+            "nunlinked                       4    4",
```

Comment:
This is the order the fields are listed in the files on my arch linux laptop, but I included tests to make sure we aren't relying on order.